src/storage/src/source/postgres.rs

```diff
@@ -129,8 +129,44 @@ impl SourceRender for PostgresSourceConnection {
         Stream<G, (usize, HealthStatusUpdate)>,
         Rc<dyn Any>,
     ) {
-        let resume_upper =
-            Antichain::from_iter(config.source_resume_upper.iter().map(MzOffset::decode_row));
+        // Get the uppers of all exports.
+        let mut export_resume_uppers: BTreeMap<_, _> = config
+            .source_resume_upper
+            .iter()
+            .map(|(id, upper)| {
+                (
+                    *id,
+                    Antichain::from_iter(upper.iter().map(MzOffset::decode_row)),
+                )
+            })
+            .collect();
+
+        // The primary source's upper is not meaningful shard, so ensure it's removed from
+        // consideration.
+        //
+        // TODO: close this upper because it can never do anything.
+        export_resume_uppers.remove(&config.id);
```

Comment:
Why do we remove this from the calculation?

Reply:
I mention removing it from the resumption frontier calculation, but a stronger motivation was ensuring it wasn't snapshot. Since I've rewritten the code a bit, I made this more clear (and also allowed it to still play into the resumption frontier calculation).